devito/nodes.py

```diff
@@ -547,16 +549,51 @@ def __repr__(self):
         return "Function[%s]<%s; %s>::\n\t%s" % (self.name, self.retval, parameters, body)
 
     @property
+    def _cargtypes(self):
+        """Return a list containing the C type of each argument, as a ``ctypes`` type."""
+        argtypes = []
+        for i in self.parameters:
+            if i.is_ScalarArgument:
+                argtypes.append(numpy_to_ctypes(i.dtype))
+            elif i.is_TensorArgument:
+                argtypes.append(ctypeslib.ndpointer(dtype=i.dtype, flags='C'))
+            else:
+                argtypes.append(ctypes.c_void_p)
+        return argtypes
+
+    @property
     def _cparameters(self):
-        """Generate arguments signature."""
-        return [v.decl for v in self.parameters]
+        """Return a list containing the Function arguments in ``cgen`` format."""
+        decls = []
+        for i in self.parameters:
+            if i.is_ScalarArgument:
+                decls.append(c.Value('const %s' % c.dtype_to_ctype(i.dtype), i.name))
+            elif i.is_TensorArgument:
+                decls.append(c.Value(c.dtype_to_ctype(i.dtype),
+                                     '*restrict %s_vec' % i.name))
+            else:
+                decls.append(c.Value('void', '*_%s' % i.name))
+        return decls
```

Comment:
Same as above. Undoes the cleanup I did earlier and brings us back to this ugly hacky if-else mess.